Rogue-Ape-Studios-XR/Secrets-of-Ignacios
# ✨ Object pooling logic

## Content

Added object pooling, changes were needed to the spellmanager script to make this work, but it has been refactored completely (sorry no separate branch, it got intertwined during the holiday 😓). Each spell now has a pool size, if the spell is added to the object pooler, it will use the pool size to initiate that amount of projectiles, if it is not enough the pool will add new projectiles during runtime. Every projectile gets used by enabling it and disabling at the correct moment.

Yes use of a singleton (because I couldn't find another way to do it) and refactored code in this branch, put me on the wall of shame if you like, I'll allow it.

## Changes
### Added
`Assets/Prefabs/Services/Services.prefab` : Was created. The gameobject needed for object pooling.
`Assets/Scripts/Puzzle/General/Spells/ProjectileSpellLogic.cs` : Was created. Was WaterSpellLogic, minor changes have been made to make it usable for all projectiles + name change.
`Assets/Scripts/Services/ObjectPooler.cs` : Was created. Object pooling logic.
`Assets/Scripts/Services/ServiceLocator.cs` : Was created. Locates the object pooling service and can locate other services (yes it is a singleton in the form of an service locator pattern).
`Assets/Scripts/Spells/Cast.cs` : Was created. Separates casting logic from spell manager scirpt.
`Assets/Prefabs/PlayerRig/PlayerRig.prefab`: Was created. Renamed the player rig prefab to be player rig + added object pooling scripts it was missing.

### Updated
`Assets/Prefabs/SpellProjectiles/Fire Spell.prefab` : Was updated / renamed. Added pool size and new projectile logic script.
`Assets/Scripts/Spells/Indicator.cs` : Was updated / renamed. Works with refactored code.
`Assets/Scripts/Spells/ScriptableObjects/Fire.asset` : Was updated / renamed. Add pool size.
`Assets/Scripts/Spells/ScriptableObjects/Water.asset` : Was updated / renamed. Add pool size.
`Assets/Scripts/Spells/Spell.cs` : Was updated / renamed. Add pool size.
`Assets/Scripts/Spells/SpellManager.cs` : Was updated / renamed. Refactored for readability and to work with object pooling.

## Testing

The feature / Bugfix can be testing by following these steps:

1. Drag player rig prefab (now correctly named) into a scene.
2. Drag Services prefab into a scene.
3. Drag services into player rig -> XR Origin (XR Rig) -> cast script -> Pooler
![image](https://github.com/user-attachments/assets/11308dc2-345c-488f-a311-f5af8598e1d4)
4. Cast spells (object pools can be found in the services object during runtime as children), you should see projectiles being enabled and disabled. When you stop player there will be unitask errors, the canceled return are from me, I don't know where the operation canceled exception is coming from (maybe becasue of the amount of tasks??), but it doesn't break anything.
![Object Pooling Small](https://github.com/user-attachments/assets/47fb93ea-3a67-4c14-9ea7-b5ef51f4a6d5)
